src/core/smp.cc

```diff
@@ -227,25 +275,29 @@ internal::memory_prefaulter::work(std::vector<memory::internal::memory_range>& r
         p1->fetch_or(0, std::memory_order_relaxed);
 #endif
     };
-    while (!_stop_request.load(std::memory_order_relaxed) && !ranges.empty()) {
-        auto& range = ranges[current_range];
-        // copy eveything into locals, or the optimizer will worry about them due
-        // to the cast below and not optimize anything
-        auto start = range.start;
-        auto end = range.end;
-        for (size_t i = 0; i < batch_size && start < end; ++i) {
+    auto populate_memory = [&] (char* start, char* end) {
+        if (populate_memory_madvise(start, end)) {
+            return;
+        }
+        while (start < end) {
             fault_in_memory(start);
             start += page_size;
         }
+    };
+    while (!_stop_request.load(std::memory_order_relaxed) && !ranges.empty()) {
+        auto& range = ranges[current_range];
+
+        auto batch_end = std::min(align_up(range.start + 1, batch_size), range.end);
+        populate_memory(range.start, batch_end);
+        range.start = batch_end;
+
         // An end-to-start scan for applications that manage two heaps that
         // grow towards each other.
-        for (size_t i = 0; i < batch_size && start < end; ++i) {
-            fault_in_memory(end - 1);
-            end -= page_size;
-        }
-        range.start = start;
-        range.end = end;
-        if (start >= end) {
+        auto batch_start = std::max(align_down(range.end - 1, batch_size), range.start);
+        populate_memory(batch_start, range.end);
+        range.end = batch_start;
+
+        if (range.start >= range.end) {
             ranges.erase(ranges.begin() + current_range);
             current_range = 0;
```

Comment:
This is pre-existing code, but I still wonder - why do we have this current_range at all? It seems it's set to 0 initially, and then always remains 0 or set to 0 again. Where is it set to anything else?

Reply:
It's incremented below.